### DISCUSSION TITLE
Error when loading Quarto html with OJS cells

### DISCUSSION THREAD

**Discussion by harrylojames** [quarto-dev/quarto-cli#5680] [Q&A]
After doing "quarto render file.qmd --to html" I am unable to double click and view the html file. Here is the error.  It does as stated work from a web server i.e. with preview. 

"The OJS runtime does not work with file:// URLs. Please use a web server to view this document."

https://quarto.org/docs/interactive/ojs/

"OJS works in any Quarto document (plain markdown as well as Jupyter and Knitr documents). Just include your code in an {ojs} executable code block. The rest of this article explains the basics of using OJS with Quarto."

This gives me the impression Quarto documents with OJS should behave the same. Am I missing something? I'm keen to produce a html with OJS that I can share and the end user click to open the file. 

Many thanks, 

Harry

**Comment by mcanouil** [comment]
Could you share a small "working" (reproducible) example to work with, _i.e._, a complete Quarto document or a Git repository? Thanks.

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Reply by harrylojames** [reply]
Sure thing

````qmd
---
title: "Reproducible Quarto Document"
format: html
---


```{ojs}
Plot.plot({
  marks: [
    Plot.barX(alphabet, {x: "frequency", y: "letter"})
  ]
})
```
````

Hope that works!

Note: adding self-contained: true or embed-resources: true removes the error but then the plot doesn't show.
```

---
title: "Reproducible Quarto Document"
format: 
    html:
        self-contained: true
---


```

**Reply by cscheid** [reply]
> "OJS works in any Quarto document (plain markdown as well as Jupyter and Knitr documents). Just include your code in an {ojs} executable code block. The rest of this article explains the basics of using OJS with Quarto."
> 
> This gives me the impression Quarto documents with OJS should behave the same. Am I missing something? I'm keen to produce a html with OJS that I can share and the end user click to open the file.

You're not missing anything. OJS _does_ work in any Quarto document. But because of features in the observablejs runtime which we have no control over, documents that use OJS need to run on the http:// or https:// protocol; they simply don't work on the file:// protocol. That's not something we can fix.

**Reply by harrylojames** [reply]
Thank you for your reply Carlos! That makes sense. 

I've created a pull request to capture this but no worries if you think this discussion sufficiently captures it for future users. 

https://github.com/quarto-dev/quarto-web/pull/739

**Reply by rafazaya** [reply]
Hello, thanks for the good question and replies you all!  My question is that can we practically use Observable cells without publishing to a hosting provider?  Is there another solution I'm missing?

**Reply by cscheid** [reply]
If you want to run locally, you can run a local web server with either `quarto preview` or `python -m http.server` or nginx or your preferred web server. If you want to share the document with other people, you can tell them to do the same thing. But by far the easiest way is to, in fact, use a hosting provider, or any of the options in `quarto publish`.

There's nothing we can do about this; it's a limitation of the Javascript features required by ObservableHQ's runtime.

**Reply by resmartiZH** [reply]
I just ran into this problem and I'm also a bit confused, since the document rendered in quarto-1.1.189 was able to run the cells in file:// without an issue, but version quarto-1.3.353 doesn't work anymore. 
For me the ability to share a plain html-file is kind of important if I want to share my reports confidentially without setting up a webhost with managed authentication and everything. So for me this is very inconvenient.

There is also a bug, with the above mentioned error message not beeing displayed in version quarto-1.3.353, but I'll make a seperate bug report for that.

**Reply by romanzenka** [reply]
I do not think it is entirely true that Observable cannot possibly run with file:// URLs. If I run quarto on a document with `format: html: embed-resources: true`, and then go back and manually delete your code that warns me that Observable cannot possibly run from file://, I get a fully functional document (works at least for the simple stuff I am doing). I rely on this functionality to deliver completely 100% self-contained documents that are guaranteed to "just work" on the client's computer.

I would recommend a check that would add the javascript file:// alerts only in case `embed-resources` was not used.

Here is a screenshot of your code running perfectly happily using file:// URL:

![image](https://github.com/quarto-dev/quarto-cli/assets/66568/73d15849-b8f7-443e-be59-36acb80e889b)

**Reply by cscheid** [reply]
It's important to distinguish between "it works for you" and "it supports all features of OJS".

**Reply by romanzenka** [reply]
@cscheid I think a good next step would be to determine the boundaries where OJS breaks. I got another example going that uses interactive controls. From what you said elsewhere, it seems that this would break those widgets that dynamically load modules in order to function.

**Reply by cscheid** [reply]
> @cscheid I think a good next step would be to determine the boundaries where OJS breaks.

Sure, go for it!

**Reply by romanzenka** [reply]
Here's a bugfix if anyone wants to try: https://github.com/romanzenka/quarto-cli/tree/bugfix/issue-6371

**Comment by romanzenka** [comment]
I have not tried it with latest Quarto changes, and as we established, it is not perfect. The problem seems to be dynamic module loading.

--Roman

**Comment by novotny1akub** [comment]
@rafazaya & @cscheid: I would like to add to the answer of cscheid on how to create a local web server that **[mongoose](https://mongoose.ws/binary/)** might be a decent option especially when sharing the OJS Quarto documents with non-tech people who are unlikely to have Python/Quarto installed and using a hosting provider is not an option. The advantage of mongoose is that it is tiny (92kB on my machine) compared to Python/Node server and does not require any additional setup. You just copy the mongoose.exe (exe is for Windows) to the same folder where you have your Quarto html document, double click it and then open http://localhost:8000 in your browser (e.g. Google Chrome). If you want to make it slightly easier for your users, you can use command line code below (again for Windows) that I put into something like `_run.bat`. Then they only need to double click the bat file and then choose the file to be displayed in the browser.

```
@echo off
start mongoose.exe
timeout /t 1 /nobreak >nul
start chrome http://localhost:8000
```

**Comment by novotny1akub** [comment]
@cscheid I also experienced something that @resmartiZH describes. After upgrading from an oldish version of Quarto where OJS worked, OJS does not seem to be working anymore in the most recent version for Quarto html with OJS opened without a web server. I agree it is a good idea to warn users that OJS may not work 100%. However, introducing changes that deliberetly break something that previously worked does not seem great. What is the aim? To encourage people to use a web server? Isn't it possible without breaking the code that may work?

**Reply by mcanouil** [reply]
You have to understand something, the policy of Posit and Quarto development team is to not break anything as much as possible, but they cannot extend this to third party tools such as Observable.
The major changes in Quarto versions regarding third parties are they (third party tools) are updated in each.

As proof of this policy, you are using an option in #7671 that does not even exist anymore in Pandoc and that is only there in Quarto to not break things.
Note that, even if it should "always" work, it's very likely Quarto will warn you in the future about using this (#7103).

**Reply by novotny1akub** [reply]
When I change `embed-resources` to `self-contained`, the issue remains the same. With Quarto 1.2.280, I can produce a fully functioning standalone html document with OJS. With the newest Quarto, I cannot.

**Reply by mcanouil** [reply]
Of course it is the same using both, it is pretty well documented, see <https://quarto.org/docs/reference/formats/html.html#rendering>.

![image](https://github.com/quarto-dev/quarto-cli/assets/8896044/b6cfd968-151a-43ec-a06e-94500893b111)
